Add an intro section to Chapter 12 (#326)
browserengineering/book: book/scheduling-and-threading.md

next: skipped
...

> To be a capable application platform, the browser must execute

Reviewer: ```suggestion
To be a capable application platform, the browser must be
```

> applications quickly and user actions immediately. To do so, the

Reviewer: ```suggestion
efficient to run application code, while also staying responsive to user actions. To do so, the
```

browser must explicitly choose which of its many tasks to prioritize
and delay unnecessary tasks until later. Such a task queue system also
allows the browser to split tasks across multiple threads, which makes

> the browser more responsive and better fit to modern hardware.

Reviewer: ```suggestion
the browser even more responsive, and a better fit to modern multi-core hardware.
```

task queues are each assigned to different threads, and each thread
chooses the most important task to work on next to keep the browser
fast and responsive. Loading pages, running scripts, and responding to
user actions all become tasks in this framework.

One of the most expensive tasks a browser does is render a web
page---style the HTML elements, construct a layout tree, compute sizes
and positions, paint it to a display list, raster the result into a
surface, and draw that surface to the screen. These rendering steps
make up the browser's basic [*rendering pipeline*][graphics-pipeline]
for the browser, and can be expensive and slow. For this reason,
modern browsers split the rendering pipeline across threads and make
sure to run the rendering pipeline only when necessary.

[graphics-pipeline]: https://en.wikipedia.org/wiki/Graphics_pipeline

Refactoring our browser to think in terms of tasks will require
significant changes throughout the browser---concurrent programming is
never easy! But these architectural changes are a key optimization
behind modern browsers, and enable many advanced features discussed in

> later chapters.

Reviewer: ```suggestion
this and later chapters.
```

next: skipped
...

To be a capable application platform, the browser must run
applications effieciently and stay responsive to user actions. To do
so, the browser must explicitly choose which of its many tasks to
prioritize and delay unnecessary tasks until later. Such a task queue
system also allows the browser to split tasks across multiple threads,
which makes the browser even more responsive and a better fit to
modern multi-core hardware.

Tasks and task queues
=====================

So far, most of the work our browser's been doing has been handling
user actions like scrolling, pressing buttons, and clicking on links.
But as our browser runs more and more sophisticated web applications,
it starts spend more and more time querying remote servers, animating
objects on the page, or prefetching information that the user may
need. This requires a change in perspective: while users are slow and
deliberative, leaving long gaps between actions for the browser to
catch up, applications can be very demanding, with a neverending queue
of tasks for the browser to do.

Modern browsers adapt to this reality by multitasking, prioritizing,

> and deduplicating work. To do so, events from the operating system are

Reviewer: ```suggestion
and de-duplicating work. To do so, events from the operating system are
```

turned into *tasks* and placed onto one of several task queues. Those
task queues are each assigned to different threads, and each thread
chooses the most important task to work on next to keep the browser
fast and responsive. Loading pages, running scripts, and responding to
user actions all become tasks in this framework.

One of the most expensive tasks a browser does is render a web
page---style the HTML elements, construct a layout tree, compute sizes
and positions, paint it to a display list, raster the result into a
surface, and draw that surface to the screen. These rendering steps

> make up the browser's basic [*rendering pipeline*][graphics-pipeline]

Reviewer: ```suggestion
make up the [*rendering pipeline*][graphics-pipeline]
```
